packages/kbn-i18n/src/core/i18n.js

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ @typedef Messages - messages tree, where leafs are translated strings
+ @type {object<string, object>}
+ @property {string} [locale] - locale of the messages
+ @property {object} [formats] - set of options to the underlying formatter
+ */
+
+import IntlMessageFormat from 'intl-messageformat';
+import IntlRelativeFormat from 'intl-relativeformat';
+import memoizeIntlConstructor from 'intl-format-cache';
+import { isString, isObject, hasValues, deepMerge } from './helper';
+import { formats as EN_FORMATS } from './formats';
+
+// Add all locale data to `IntlMessageFormat`.
+// TODO: Use dynamic import for asynchronous loading of specific locale data
+import './locales';
+
+const EN_LOCALE = 'en';
+const LOCALE_DELIMITER = '-';
+const messages = {};
+const getMessageFormat = memoizeIntlConstructor(IntlMessageFormat);
+
+let defaultLocale = EN_LOCALE;
+let currentLocale = EN_LOCALE;
+let formats = EN_FORMATS;
+
+IntlMessageFormat.defaultLocale = defaultLocale;
+IntlRelativeFormat.defaultLocale = defaultLocale;
+
+/**
+ * Returns message by the given message id.
+ * @param {Messages} messages - messages tree
+ * @param {string} id - path to the message that consists of properties
+ * names separated by dots
+ * @returns {string} message - translated message from messages tree
+ * @example
+ * getMessageById({ a: { b: { c: 'test' } } }, 'a.b.c'); // => 'test'
+ */
+function getMessageById(messages, id) {
+  return id.split('.').reduce((val, key) => (val ? val[key] : null), messages);
+}
+
+/**
+ * Normalizes locale to make it consistent with IntlMessageFormat locales
+ * @param {string} locale
+ * @returns {string} normalizedLocale
+ */
+function normalizeLocale(locale) {
+  return locale.toLowerCase().replace('_', LOCALE_DELIMITER);
+}
+
+/**
+ * Provides a way to register translations with the engine
+ * @param {Messages} newMessages
+ * @param {string} [locale = messages.locale]
+ */
+export function addMessages(newMessages = {}, locale = newMessages.locale) {
+  if (!locale) {
+    throw new Error('[I18n] A `locale` must be provided to add messages.');
+  } else {
```

Comment:
nit: There's no reason to wrap the rest of this function in an else here since the `throw` exits this function context if it's executed.